### ISSUE TITLE
Memory allocation issues

### ISSUE THREAD

**Issue by kparichay** [nnstreamer/nntrainer#974]
1. Memory allocation for a model is done again and again per run, without allocating memory from the previous run. - #976
2. Memory deallocating at the destruction of model is not done properly. - #976
3. Memory allocation must be done again at setting the batch size, and previous memory must be deallocated before next memory is allocated. - #976
4. Memory must be freed at the end of the training run.
5. Memory must not be freed at the end of inference run, but this memory must be freed before starting another training run. Also create interface to free inference run memory.
6. None of these will free weight memory for the model. This will be freed at the destruction of the model only.

**Comment by kparichay** [comment]
Also, add unittests for the manager, initialize, and allocation.

**Comment by kparichay** [comment]
Create another class in Manager to manage states of weights/grads/inouts/deriv as suggested by @zhoonit. This will be done in a new PR soon.

**Comment by zhoonit** [comment]
Could you consider below as discussed offline a few days ago?
1. restricting the ordering of `initialize - allocate - deallocate - deinitialize` (else hard error).

**Comment by kparichay** [comment]
Yes, this will be done soon. As we discussed offline earlier, 
1. the manager code will be more modularized (if possible), and check the ordering of `initialize - (allocate - deallocate) - deinitialize`
2. rename functions in weights and var_grads from `initialize` to what they actually do.

**Comment by jijoongmoon** [comment]
You can close this if it is done.

**Comment by kparichay** [comment]
Once #1001 is merged, https://github.com/nnstreamer/nntrainer/issues/974#issuecomment-797207164 needs to be addressed, which is about refactoring manager and weights.
As this is more towards refactoring and not for addressing any allocation issues, I have moved its tracking to #986.
This will close once #1001 is merged.
